data_diff/dbt.py

```diff
@@ -0,0 +1,313 @@
+                )
+
+        rich.print("Diffs Complete!")
+
+    @staticmethod
+    def _get_diff_vars(dbt_parser, config_prod_database, config_prod_schema, model, datasource_id) -> DiffVars:
+        dev_database = model.database
+        dev_schema = model.schema_
+        primary_keys = dbt_parser.get_primary_keys(model)
+
+        if config_prod_database is None:
+            prod_database = dev_database
+        else:
+            prod_database = config_prod_database
+
+        if config_prod_schema is None:
+            prod_schema = dev_schema
+        else:
+            prod_schema = config_prod_schema
+
+        if dbt_parser.requires_upper:
+            dev_qualified_list = [x.upper() for x in [dev_database, dev_schema, model.name]]
+            prod_qualified_list = [x.upper() for x in [prod_database, prod_schema, model.name]]
+            primary_keys = [x.upper() for x in primary_keys]
+        else:
+            dev_qualified_list = [dev_database, dev_schema, model.name]
+            prod_qualified_list = [prod_database, prod_schema, model.name]
+        return DiffVars(dev_qualified_list, prod_qualified_list, primary_keys, datasource_id, dbt_parser.connection)
+
+    @staticmethod
+    def _local_diff(diff_vars: DiffVars):
+        column_diffs_str = ""
+        dev_qualified_string = ".".join(diff_vars.dev_path)
+        prod_qualified_string = ".".join(diff_vars.prod_path)
+        primary_key = diff_vars.primary_keys[0]
+
+        table1 = connect_to_table(diff_vars.connection, dev_qualified_string, primary_key)
+        table2 = connect_to_table(diff_vars.connection, prod_qualified_string, primary_key)
+
+        table1_columns = list(table1.get_schema())
+        try:
+            table2_columns = list(table2.get_schema())
+        # TODO add a unit test for scenario
+        except Exception as e:
+            logging.debug(e)
+            rich.print(
+                "[red]"
+                + dev_qualified_string
+                + " <> "
+                + prod_qualified_string
+                + "[/] \n"
+                + column_diffs_str
+                + "[green]New model or no access to prod table.[/] \n"
+            )
+            return
+
+        mutual_set = set(table1_columns) & set(table2_columns)
+        table1_set_diff = list(set(table1_columns) - set(table2_columns))
+        table2_set_diff = list(set(table2_columns) - set(table1_columns))
+
+        if len(table1_set_diff) > 0:
+            column_diffs_str += "Columns exclusive to table A: " + str(table1_set_diff) + "\n"
+
+        if len(table2_set_diff) > 0:
+            column_diffs_str += "Columns exclusive to table B: " + str(table2_set_diff) + "\n"
+
+        mutual_set.discard(primary_key)
+        extra_columns = tuple(mutual_set)
+
+        diff = diff_tables(table1, table2, threaded=True, algorithm=Algorithm.JOINDIFF, extra_columns=extra_columns)
+
+        if len(list(diff)) > 0:
```

Comment:
Mind that the difference can be huge (e.g. thousands or millions of rows). This place (`list(diff)`) will try to materialize them all into memory before doing a check.

Perhaps, you want to only check if it yields at least one item with `next(diff)`. If it is empty, it will raise `StopIteration`. Alternatively, `next(diff, None)` can be used to return `None` (or any other placeholder) if the diff is empty, the first item otherwise. In that case, it will get one diff item, abandon the object, and it will be garbage-collected almost instantly.

Reply:
that materialization is needed in order to have stats to print with `diff.get_stats_string()` and will occur if I call that method regardless

see:
https://github.com/datafold/data-diff/blob/2d624d651991826e2bf652fd8fb85e504c6dd5d8/data_diff/diff_tables.py#L104

Reply:
Ah, I see. That is a place for a potential problem then. It should be iterating over `self` until depleted, not constructing a `list()` and immediately discarding it. 

Nevertheless, this is out of the scope of this PR and should be fixed separately. Thanks for clarifying.